[FIX] website_sale_stock : show correct available qty in product page

Mtaylorr opened the pull request:
When you add items in the cart it always show the number of available left in stock of the previous state for example : 
1- cart is empty it shows 20 items
2- add one into cart -  it shows 20 items left in stock  (should be 19) 
3- add another one into cart - it show 19 items left in stock (should be 18)

Steps to reproduce the error :
1 - Go to one of the product and make sure that it has some available quantity 
2 - Go to sale configuration on the product and remove "out-of-stock : continue selling" and add  show available qty only if below 5000 for example 
3 - Go the web page of the product and try to add some items into the cart you will see that the stock message is wrong

The origin of the problem is that there is 2 transactions being made to the database at the same time , one to update the cart_quantity in the sale_order and one to retrieve that quantity. Since the isolation level doesn't allow to read uncommitted changes , the read request will get the old value of cart_quantity which explains why there is a delay between the message and the real quantity. I added the timeout so it will executes the read once the stack is empty and we are sure that the updates are committed.

opw-3341377

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/124164).

nle-odoo reviewed (approved):
Hi

Testing in local with only website_sale_stock and dependencies installed, I can't reproduce the issue and the `update_json` query is finished before the `get_combination_info_website`.

Testing on runbot, it is easy to reproduce. Checking in more details, it seems that the issue is that website_sale_product_configurator breaks the chain of deferred, so we don't wait for update_json before updating get_combination_info_website.

So the current fix doesn't seem to be the best, it might work but depending on network latency there could still be an issue.

I'm not too familiar with this code, but a change like this seems to fix the issue:

```diff
diff --git a/addons/website_sale_product_configurator/static/src/js/website_sale_options.js b/addons/website_sale_product_configurator/static/src/js/website_sale_options.js
index d95ea5bcfe3..606af874838 100644
--- a/addons/website_sale_product_configurator/static/src/js/website_sale_options.js
+++ b/addons/website_sale_product_configurator/static/src/js/website_sale_options.js
@@ -42,10 +42,11 @@ publicWidget.registry.WebsiteSale.include({
      * @override
      */
     _submitForm() {
+        var ret = this._super(...arguments);
         if (this.optionalProductsModal && this.stayOnPageOption) {
-            this.optionalProductsModal._openedResolver();
+            ret.then(this.optionalProductsModal._openedResolver);
         }
-        return this._super(...arguments);
+        return ret;
     },
     /**
      * Update web shop base form quantity
      *
```
because it delays the modal opened status to when the update_json is finished, and the code is waiting for the modal opened status (instead of that the update_json is finished) in this override:

https://github.com/odoo/odoo/blob/ea1050c81b1a2563a471b425c7f9c9b6af3404b9/addons/website_sale_product_configurator/static/src/js/website_sale_options.js#L36

Mtaylorr commented:
The fix seems to work only when there are no options for the product or the setting for optional products is inactve, in case there are optional products , But in case the there are optional products, the update of the message happens before we even confirm the update, because the promise will wait until the modal is opened and not the update have been made , so the flow is like this : wait for the model to open => promise resolved => then call this._getCombinationInfo(ev); => the user then can add some more items and finally confirm it => update cart, so the cart is updated but the message of the stock remains the same

Mtaylorr commented:
> ok, then this use case might need a more complicated fix, maybe we could use the add_to_cart event, and send this event if there is no optional product modal

If the optional products setting is off , it works fine (no need for any fix), 
If it has no optional products , the fix you proposed works, 
The main problem is when the model opens and then closes, the message doesn't change.

I think a solution is to write a function similar to _getCombinationInfo (called here : https://github.com/odoo/odoo/blob/9a3f5c042fa6687a885f4dcb70efc9558f2da543/addons/website_sale_stock/static/src/js/variant_mixin.js#L99 and its code is here https://github.com/odoo/odoo/blob/9a3f5c042fa6687a885f4dcb70efc9558f2da543/addons/sale/static/src/js/variant_mixin.js#L51) to update the message or we change the current one because this one accept an event as input, what do you think ?

Mtaylorr commented:
@nle-odoo I added some changes, something like this would work but I don't know if it is really a good change because I changed many files and "some logic" instead of sending the event I send the element. I saw that the event is only used for the target so I changed all the params to target in this way I can call the function without the need for an event which what I did in `_onModalSubmit`

Feyensv commented on addons/website_sale_product_configurator/static/src/js/website_sale_options.js:
```diff
@@ -113,6 +114,8 @@ publicWidget.registry.WebsiteSale.include({
                         const $quantity = $(".my_cart_quantity");
                         $quantity.parent().parent().removeClass('d-none');
                         $quantity.text(quantity).hide().fadeIn(600);
+                    }).then(()=>{
+                        this._getCombinationInfo($("#add_to_cart_wrap"));
```
Wouldn't it be easier to simply refresh the page instead of modifying the `_getCombinationInfo` API (which seems big for stable bugfix) ?

nle-odoo replied:
we could also just give `$.Event('click', {target: $("#add_to_cart_wrap")})` and not change the signature

Feyensv reviewed (commented):
Not sure we want to change that much code for this little case :thinking: (even more considering it's used in backend and frontend).

Also, could you please add a test covering this broken flow ? :pray: 
As we made (and are going to make) a lot of changes in the configurator flows, it's better if those advanced problems are covered to make sure we do not break them again.

Thanks

nle-odoo commented:
There is still the setTimeout that seemed to me a bad way (because it's not fixing the issue, and the problem might still happen depending on network latency) to fix the issue. 

For the current fix, another way might be to use a promise when needed but it's still a lot of changes:

https://gist.github.com/Etana/5c27ffa1aa0ac886c9b924ea59fbf194

Feyensv commented on addons/website_sale_product_configurator/static/tests/tours/website_sale_stock_message_after_close.js:
```diff
@@ -0,0 +1,61 @@
+odoo.define('website_sale.website_sale_stock_message_after_close', function (require) {
```
The two tours must be defined in `website_sale_stock_product_configurator`, they fail when `website_sale_stock` is not installed.

Mtaylorr commented:
Hello @Feyensv , can you reveiw the last changed I added, thanks

Mtaylorr commented on addons/website_sale_stock_product_configurator/tests/test_website_sale_stock_product_configurator.py:
```diff
@@ -0,0 +1,50 @@
+class TestWebsiteSaleStockProductConfigurator(TestProductConfiguratorCommon, HttpCaseWithUserPortal):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
```
I removed it , I added it at first and then I didn't use it , Thanks.
